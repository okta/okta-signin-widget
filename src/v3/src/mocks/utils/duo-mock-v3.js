module.exports = {
  init: (options) => {
    // Load a mock iframe
    // and add an event listener to continue with authentication
    options.iframe.src = 'duo-iframe.html';
    options.iframe.onload = () => {
      var innerDoc = options.iframe.contentDocument || options.iframe.contentWindow.document;
      var duoMockLink = innerDoc.getElementById('duoVerifyLink');
      duoMockLink.addEventListener('click', () => {
        options.post_action('successDuoAuth');
      }, false);
    };
  }
};
