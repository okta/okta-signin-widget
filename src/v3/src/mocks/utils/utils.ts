/*
 * Copyright (c) 2022-present, Okta, Inc. and/or its affiliates. All rights reserved.
 * The Okta software accompanied by this notice is provided pursuant to the Apache License, Version 2.0 (the "License.")
 *
 * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0.
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *
 * See the License for the specific language governing permissions and limitations under the License.
 */

import { IdxStatus, IdxTransaction } from '@okta/okta-auth-js';
import { IDX_STEP } from 'src/constants';
import { DataSchemaBag, FormBag, UISchemaLayoutType } from 'src/types';

export const getStubTransaction = (status: IdxStatus = IdxStatus.PENDING): IdxTransaction => ({
  status,
  proceed: jest.fn(),
  neededToProceed: [],
  rawIdxState: {
    version: '',
    stateHandle: '',
  },
  actions: {},
  context: {
    version: '',
    stateHandle: '',
    expiresAt: '',
    intent: '',
    currentAuthenticator: {
      type: '',
      value: {
        displayName: '',
        id: '',
        methods: [],
        key: '',
        type: '',
      },
    },
    authenticators: {
      type: '',
      value: [],
    },
    authenticatorEnrollments: {
      type: '',
      value: [],
    },
    enrollmentAuthenticator: {
      type: '',
      value: {
        displayName: '',
        id: '',
        methods: [],
        key: '',
        type: '',
      },
    },
    currentAuthenticatorEnrollment: {
      type: '',
      value: {
        displayName: '',
        id: '',
        methods: [],
        key: '',
        type: '',
      },
    },
    user: {
      type: '',
      value: {},
    },
    app: {
      type: '',
      value: {},
    },
  },
});

export const getStubTransactionWithNextStep = (stepName?: string): IdxTransaction => {
  const transaction = getStubTransaction();

  return { ...transaction, nextStep: { name: stepName ?? '' } };
};

export const getStubFormBag = (
  step?: string,
  orientation?: UISchemaLayoutType,
): FormBag => ({
  dataSchema: {
    submit: { step: step || IDX_STEP.CHALLENGE_AUTHENTICATOR },
    fieldsToTrim: [],
    fieldsToValidate: [],
    fieldsToExclude: () => ([]),
  } as DataSchemaBag,
  schema: {},
  uischema: {
    type: orientation || UISchemaLayoutType.VERTICAL,
    elements: [],
  },
  data: {},
});

export const getMockCreateCredentialsResponse = (): PublicKeyCredential => (
  {
    id: 'test',
    type: 'test',
    authenticatorAttachment: null,
    rawId: new ArrayBuffer(10),
    getClientExtensionResults: jest.fn(),
    response: {
      clientDataJSON: new ArrayBuffer(10),
      attestationObject: new ArrayBuffer(10),
    } as AuthenticatorAttestationResponse,
  } as PublicKeyCredential
);
