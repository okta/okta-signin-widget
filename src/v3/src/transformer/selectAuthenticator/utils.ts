/*
 * Copyright (c) 2022-present, Okta, Inc. and/or its affiliates. All rights reserved.
 * The Okta software accompanied by this notice is provided pursuant to the Apache License, Version 2.0 (the "License.")
 *
 * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0.
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *
 * See the License for the specific language governing permissions and limitations under the License.
 */

import { Input } from '@okta/okta-auth-js';
import { IdxOption } from '@okta/okta-auth-js/types/lib/idx/types/idx-js';

import {
  AUTHENTICATOR_ALLOWED_FOR_OPTIONS,
  AUTHENTICATOR_ENROLLMENT_DESCR_KEY_MAP,
  AUTHENTICATOR_KEY,
} from '../../constants';
import { ActionParams, AuthenticatorButtonElement, ButtonType } from '../../types';
import { loc } from '../../util';

export const getOptionValue = (
  inputs: Input[],
  key: string,
): Input | undefined => inputs?.find(
  ({ name }) => name === key,
);

const getAuthenticatorDataSeVal = (authenticatorKey: string, methodType?: string): string => {
  if (authenticatorKey) {
    const method = methodType ? `-${methodType}` : '';
    return `${authenticatorKey}${method}`;
  }
  return '';
};

const reorderAuthenticatorButtons = (
  authButtons: AuthenticatorButtonElement[],
  deviceKnown?: boolean,
): AuthenticatorButtonElement[] => {
  if (authButtons.length <= 1) {
    return authButtons;
  }
  const fastpassAuthenticator = authButtons.find(
    (button) => button.options.actionParams?.['authenticator.methodType'] === 'signed_nonce',
  );
  if (!fastpassAuthenticator) {
    return authButtons;
  }

  const updatedAuthenticatorBtns = authButtons.filter(
    (button) => button.options.actionParams?.['authenticator.methodType'] !== 'signed_nonce',
  );

  // Re-arrange fastpass in options based on deviceKnown
  // If deviceKnown is set, set fastpass as the first option in the list
  // otherwise, place it as the last item in the list of OV options
  if (deviceKnown) {
    updatedAuthenticatorBtns.unshift(fastpassAuthenticator);
  } else {
    updatedAuthenticatorBtns.push(fastpassAuthenticator);
  }

  return updatedAuthenticatorBtns;
};

const buildOktaVerifyOptions = (
  options: IdxOption[],
  step: string,
  isEnroll?: boolean,
): AuthenticatorButtonElement[] => {
  const ovRemediation = options.find((option) => option.relatesTo?.key === AUTHENTICATOR_KEY.OV);
  const id = (ovRemediation?.value as Input[])?.find(({ name }) => name === 'id')?.value;
  const methodType = (ovRemediation?.value as Input[])?.find(({ name }) => name === 'methodType');
  if (!methodType?.options?.length) {
    return [];
  }

  return methodType.options.map((option: IdxOption, index: number) => {
    const authenticatorButton: AuthenticatorButtonElement = {
      type: 'AuthenticatorButton',
      label: option.label,
      id: `auth_btn_${AUTHENTICATOR_KEY.OV}_${option.value || id}`,
      options: {
        type: ButtonType.BUTTON,
        key: AUTHENTICATOR_KEY.OV,
        ctaLabel: isEnroll
          ? loc('oie.enroll.authenticator.button.text', 'login')
          : loc('oie.verify.authenticator.button.text', 'login'),
        description: isEnroll
          ? loc(AUTHENTICATOR_ENROLLMENT_DESCR_KEY_MAP[AUTHENTICATOR_KEY.OV], 'login')
          : loc('oie.okta_verify.label', 'login'),
        actionParams: {
          'authenticator.methodType': option.value,
          'authenticator.id': id,
        } as ActionParams,
        step,
        includeData: true,
        includeImmutableData: false,
        dataSe: getAuthenticatorDataSeVal(AUTHENTICATOR_KEY.OV, option.value as string),
        iconName: option.value === 'totp' ? `oktaVerify_${index}` : `oktaVerifyPush_${index}`,
        iconDescr: option.value === 'totp'
          ? loc('factor.totpSoft.description', 'login')
          : loc('factor.push.description', 'login'),
      },
    };
    return authenticatorButton;
  });
};

const getAuthenticatorDescriptionParams = (
  option: IdxOption,
  authenticatorKey: string,
  isEnroll?: boolean,
): string[] | undefined => {
  const authenticatorDescrWithParams = [
    AUTHENTICATOR_KEY.ON_PREM,
    AUTHENTICATOR_KEY.IDP,
    AUTHENTICATOR_KEY.CUSTOM_APP,
    AUTHENTICATOR_KEY.SYMANTEC_VIP,
  ];
  if (!isEnroll || !authenticatorDescrWithParams.includes(authenticatorKey)) {
    return undefined;
  }

  switch (authenticatorKey) {
    case AUTHENTICATOR_KEY.ON_PREM: {
      const vendorName = option.relatesTo?.displayName
        || loc('oie.on_prem.authenticator.default.vendorName', 'login');
      return [vendorName];
    }
    case AUTHENTICATOR_KEY.IDP: {
      const idpName = option.relatesTo?.displayName || '';
      return [idpName];
    }
    case AUTHENTICATOR_KEY.CUSTOM_APP: {
      const customAppName = option.label || '';
      return [customAppName];
    }
    case AUTHENTICATOR_KEY.SYMANTEC_VIP: {
      const appName = option.relatesTo?.displayName || '';
      return [appName];
    }
    default:
      return undefined;
  }
};

const getAuthenticatorDescription = (
  option: IdxOption,
  authenticatorKey: string,
  isEnroll?: boolean,
): string | undefined => {
  if (!authenticatorKey) {
    return undefined;
  }
  const descrParams = getAuthenticatorDescriptionParams(
    option,
    authenticatorKey,
    isEnroll,
  );
  if (isEnroll) {
    return loc(AUTHENTICATOR_ENROLLMENT_DESCR_KEY_MAP[authenticatorKey], 'login', descrParams);
  }

  switch (authenticatorKey) {
    case AUTHENTICATOR_KEY.PHONE:
      return option.relatesTo?.profile?.phoneNumber as string || undefined;
    case AUTHENTICATOR_KEY.EMAIL:
      return option.relatesTo?.profile?.email as string || undefined;
    case AUTHENTICATOR_KEY.CUSTOM_APP:
      return option.relatesTo?.displayName as string || undefined;
    case AUTHENTICATOR_KEY.OV:
      return loc('oie.okta_verify.label', 'login');
    default:
      return undefined;
  }
};

const getNickname = (
  option: IdxOption,
  authenticatorKey: string,
  isEnroll?: boolean,
): string | undefined => {
  if (!authenticatorKey || isEnroll) {
    return undefined;
  }

  if (authenticatorKey === AUTHENTICATOR_KEY.PHONE) {
    // @ts-expect-error OKTA-661650 nickname missing from IdxAuthenticator
    return option.relatesTo?.nickname;
  }
  return undefined;
};

const getUsageDescription = (option: IdxOption): string | undefined => {
  // @ts-ignore IdxAuthenticator missing allowedFor property
  const { allowedFor } = option.relatesTo;
  switch (allowedFor) {
    case AUTHENTICATOR_ALLOWED_FOR_OPTIONS.ANY:
      return loc('oie.enroll.authenticator.usage.text.access.recovery', 'login');
    case AUTHENTICATOR_ALLOWED_FOR_OPTIONS.RECOVERY:
      return loc('oie.enroll.authenticator.usage.text.recovery', 'login');
    case AUTHENTICATOR_ALLOWED_FOR_OPTIONS.SSO:
      return loc('oie.enroll.authenticator.usage.text.access', 'login');
    default:
      return undefined;
  }
};

const formatAuthenticatorOptions = (
  options: IdxOption[],
  step: string,
  isEnroll?: boolean,
): AuthenticatorButtonElement[] => {
  const authenticatorOptionSet = new Set<string>();
  return options
    .filter((option: IdxOption) => {
      if (isEnroll) {
        return true;
      }
      // If webauthn enrollments > 1 just show one entry with a generic namne (first)
      // so user doesnt have to select which one to pick. eg)
      // If there is yubikey5 and another unknown u2f key, user cannot identify that easily.
      // We need to do this at least  until users can give authenticator enrollments custom names.
      let isDup = false;
      const authenticatorKey = option.relatesTo?.key as string;
      if (authenticatorKey === AUTHENTICATOR_KEY.WEBAUTHN) {
        isDup = authenticatorOptionSet.has(authenticatorKey);
        authenticatorOptionSet.add(authenticatorKey);
      } else if (authenticatorKey === AUTHENTICATOR_KEY.CUSTOM_APP) {
        const id = getOptionValue(option.value as Input[], 'id')?.value as string;
        isDup = authenticatorOptionSet.has(id);
        authenticatorOptionSet.add(id);
      }
      return !isDup;
    })
    .map((option: IdxOption, index: number) => {
      const authenticatorKey = option.relatesTo?.key as string;
      const id = getOptionValue(option.value as Input[], 'id')?.value;
      const methodType = getOptionValue(option.value as Input[], 'methodType')?.value;
      const enrollmentId = getOptionValue(option.value as Input[], 'enrollmentId')?.value;
      const AUTHENTICATORS_WITH_METHOD_TYPE = [
        AUTHENTICATOR_KEY.ON_PREM,
        AUTHENTICATOR_KEY.OV,
        AUTHENTICATOR_KEY.RSA,
      ];
      const AUTHENTICATORS_WITH_NO_TRANSLATE_CLASS = [
        AUTHENTICATOR_KEY.PHONE,
        AUTHENTICATOR_KEY.CUSTOM_APP,
      ];
      const AUTHENTICATORS_WITH_LTR_DESCRIPTION = [
        AUTHENTICATOR_KEY.PHONE,
      ];
      const authenticator = option.relatesTo;

      return {
        type: 'AuthenticatorButton',
        label: option.label,
        id: `auth_btn_${authenticatorKey}_${enrollmentId || id}`,
        noTranslate: !isEnroll && AUTHENTICATORS_WITH_NO_TRANSLATE_CLASS.includes(authenticatorKey),
        dir: !isEnroll && AUTHENTICATORS_WITH_LTR_DESCRIPTION.includes(authenticatorKey) ? 'ltr' : undefined,
        options: {
          type: ButtonType.BUTTON,
          key: authenticatorKey,
          authenticator,
          ctaLabel: isEnroll
            ? loc('oie.enroll.authenticator.button.text', 'login')
            : loc('oie.verify.authenticator.button.text', 'login'),
          description: getAuthenticatorDescription(
            option,
            authenticatorKey,
            isEnroll,
          ),
          nickname: getNickname(option, authenticatorKey, isEnroll),
          usageDescription: isEnroll && getUsageDescription(option),
          // @ts-ignore logoUri missing from interface
          logoUri: authenticator.logoUri,
          actionParams: {
            'authenticator.id': id,
            'authenticator.methodType': AUTHENTICATORS_WITH_METHOD_TYPE.includes(authenticatorKey)
              ? methodType
              : undefined,
            'authenticator.enrollmentId': enrollmentId,
          },
          step,
          includeData: true,
          includeImmutableData: false,
          dataSe: getAuthenticatorDataSeVal(
            authenticatorKey,
            AUTHENTICATORS_WITH_METHOD_TYPE.includes(authenticatorKey) && typeof methodType === 'string'
              ? methodType
              : undefined,
          ),
          iconName: `${authenticatorKey}_${index}`,
        },
      } as AuthenticatorButtonElement;
    });
};

const getAuthenticatorButtonElements = (
  options: IdxOption[],
  step: string,
  isEnroll?: boolean,
): AuthenticatorButtonElement[] => {
  const formattedOptions = formatAuthenticatorOptions(options, step, isEnroll);

  // appending OV options back to its original spot
  const ovOptions = buildOktaVerifyOptions(options, step, isEnroll);
  if (ovOptions.length && options?.length) {
    const ovIndex = options.findIndex(({ relatesTo }) => relatesTo?.key === AUTHENTICATOR_KEY.OV);
    formattedOptions.splice(ovIndex, 1, ...ovOptions);
  }

  const ovRemediation = options.find((option) => option.relatesTo?.key === AUTHENTICATOR_KEY.OV);
  const methodType = (ovRemediation?.value as Input[])?.find(({ name }) => name === 'methodType');
  // Only re-order auth buttons when options contains OV with signed_nonce method type
  if (!methodType?.options?.find((option: IdxOption) => option.value === 'signed_nonce')) {
    return formattedOptions;
  }
  const deviceKnown = ovRemediation?.relatesTo?.deviceKnown;
  return reorderAuthenticatorButtons(formattedOptions, deviceKnown);
};

export const getAppAuthenticatorMethodButtonElements = (
  authenticator: Input,
  step: string,
  authKey = AUTHENTICATOR_KEY.OV,
  deviceKnown?: boolean,
): AuthenticatorButtonElement[] => {
  const id = (authenticator.value as Input[])?.find(({ name }) => name === 'id')?.value as string;
  const methodType = (authenticator.value as Input[])?.find(({ name }) => name === 'methodType');
  if (!methodType?.options?.length) {
    return [];
  }

  const authButtons = methodType.options.map((option, index) => ({
    type: 'AuthenticatorButton',
    label: option.label,
    id: `auth_btn_${authKey}_${option.value as string}`,
    noTranslate: authKey === AUTHENTICATOR_KEY.CUSTOM_APP,
    options: {
      type: ButtonType.BUTTON,
      key: authKey,
      ctaLabel: loc('oie.verify.authenticator.button.text', 'login'),
      actionParams: {
        'authenticator.id': id,
        'authenticator.methodType': (option.value as string),
      },
      description: getAuthenticatorDescription(
        option,
        authKey,
        false,
      ),
      dataSe: getAuthenticatorDataSeVal(
        authKey,
        option.value as string,
      ),
      iconName: `${authKey}_${index}`,
      step,
      includeData: true,
      includeImmutableData: false,
    },
  })) as AuthenticatorButtonElement[];

  return reorderAuthenticatorButtons(authButtons, deviceKnown);
};

export const isOnlyPushWithAutoChallenge = (
  inputs?: Input[],
): boolean => {
  const methodType = inputs?.find(({ name }) => name === 'methodType');
  const autoChallenge = inputs?.find(({ name }) => name === 'autoChallenge');

  return typeof autoChallenge !== 'undefined'
    && methodType?.options?.length === 1
    && methodType.options[0].value === 'push';
};

export const getAuthenticatorVerifyButtonElements = (
  authenticatorOptions: IdxOption[],
  step: string,
):AuthenticatorButtonElement[] => getAuthenticatorButtonElements(
  authenticatorOptions,
  step,
);

export const getAuthenticatorEnrollButtonElements = (
  authenticatorOptions: IdxOption[],
  step: string,
): AuthenticatorButtonElement[] => getAuthenticatorButtonElements(
  authenticatorOptions,
  step,
  true,
);
