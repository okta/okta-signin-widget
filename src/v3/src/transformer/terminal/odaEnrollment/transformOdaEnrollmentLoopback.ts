/*
 * Copyright (c) 2022-present, Okta, Inc. and/or its affiliates. All rights reserved.
 * The Okta software accompanied by this notice is provided pursuant to the Apache License, Version 2.0 (the "License.")
 *
 * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0.
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *
 * See the License for the specific language governing permissions and limitations under the License.
 */

import { CHALLENGE_METHOD, OKTA_VERIFY_APP_URL } from '../../../constants';
import {
  ButtonElement,
  ButtonType,
  DescriptionElement,
  IdxStepTransformer,
  ListElement,
  TitleElement,
  UISchemaLayout,
  UISchemaLayoutType,
} from '../../../types';
import { copyToClipboard, loc } from '../../../util';

export const transformOdaEnrollmentLoopback: IdxStepTransformer = ({ formBag, transaction }) => {
  // @ts-expect-error Property 'deviceEnrollment' does not exist on type 'IdxContext' ts(2339)
  const deviceEnrollment = transaction.context?.deviceEnrollment?.value;

  const {
    challengeMethod,
    signInUrl,
    platform: rawPlatform = '',
  } = deviceEnrollment;
  const platform: string = rawPlatform.toLowerCase();
  const isIOS = platform === 'ios';
  const isAndroid = platform === 'android';

  formBag.uischema.elements.push({
    type: 'Title',
    options: { content: loc('enroll.title.oda', 'login') },
  } as TitleElement);

  formBag.uischema.elements.push({
    type: 'Description',
    contentType: 'subtitle',
    options: { content: loc('enroll.explanation.p1', 'login') },
  } as DescriptionElement);

  const listItems: (string | UISchemaLayout)[] = [];

  if (isIOS) {
    listItems.push({
      type: UISchemaLayoutType.VERTICAL,
      elements: [
        {
          type: 'Description',
          noMargin: true,
          options: {
            content: loc('enroll.mdm.step.copyLink', 'login'),
          },
        } as DescriptionElement,
        {
          type: 'Button',
          label: loc('enroll.mdm.copyLink'),
          options: {
            step: '',
            type: ButtonType.BUTTON,
            variant: 'secondary',
            onClick: () => copyToClipboard(OKTA_VERIFY_APP_URL.IOS),
          },
        } as ButtonElement,
      ],
    } as UISchemaLayout,
    {
      type: UISchemaLayoutType.VERTICAL,
      elements: [
        {
          type: 'Description',
          noMargin: true,
          options: {
            content: loc('enroll.mdm.step.pasteLink', 'login'),
          },
        } as DescriptionElement,
      ],
    } as UISchemaLayout,
    {
      type: UISchemaLayoutType.VERTICAL,
      elements: [
        {
          type: 'Description',
          noMargin: true,
          options: {
            content: loc('enroll.oda.step3', 'login'),
          },
        } as DescriptionElement,
      ],
    } as UISchemaLayout);
  } else if (isAndroid && challengeMethod === CHALLENGE_METHOD.LOOPBACK) {
    listItems.push({
      type: UISchemaLayoutType.VERTICAL,
      elements: [
        {
          type: 'Description',
          noMargin: true,
          options: {
            content: loc('enroll.oda.android.step1', 'login'),
          },
        } as DescriptionElement,
      ],
    });
  }

  listItems.push({
    type: UISchemaLayoutType.VERTICAL,
    elements: [
      {
        type: 'Description',
        noMargin: true,
        options: {
          content: loc('enroll.oda.step1', 'login'),
        },
      } as DescriptionElement,
    ],
  } as UISchemaLayout,
  {
    type: UISchemaLayoutType.VERTICAL,
    elements: [
      {
        type: 'Description',
        noMargin: true,
        options: {
          content: loc('enroll.oda.step2', 'login'),
        },
      } as DescriptionElement,
      {
        type: 'Description',
        noMargin: true,
        options: {
          content: signInUrl,
        },
      } as DescriptionElement,
      {
        type: 'Button',
        label: loc('enroll.oda.org.copyLink', 'login'),
        options: {
          step: '',
          type: ButtonType.BUTTON,
          variant: 'secondary',
          onClick: () => copyToClipboard(deviceEnrollment?.signInUrl),
        },
      } as ButtonElement,
    ],
  } as UISchemaLayout,
  {
    type: UISchemaLayoutType.VERTICAL,
    elements: [
      {
        type: 'Description',
        noMargin: true,
        options: {
          content: loc('enroll.oda.step6', 'login'),
        },
      } as DescriptionElement,
    ],
  } as UISchemaLayout);

  formBag.uischema.elements.push({
    type: 'List',
    noMargin: true,
    options: {
      type: 'ordered',
      items: listItems,
    },
  } as ListElement);

  return formBag;
};
