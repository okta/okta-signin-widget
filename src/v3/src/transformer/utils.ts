/*
 * Copyright (c) 2022-present, Okta, Inc. and/or its affiliates. All rights reserved.
 * The Okta software accompanied by this notice is provided pursuant to the Apache License, Version 2.0 (the "License.")
 *
 * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0.
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *
 * See the License for the specific language governing permissions and limitations under the License.
 */

import {
  FieldElement,
  FormBag,
  UISchemaElement,
  UISchemaLayoutType,
} from '../types';

export const createForm = (): FormBag => ({
  schema: {},
  uischema: {
    type: UISchemaLayoutType.VERTICAL,
    elements: [],
  },
  data: {},
  dataSchema: {
    fieldsToValidate: [] as string[],
  } as FormBag['dataSchema'],
});

export const removeUIElementWithName = (
  name: string,
  elements: UISchemaElement[],
): UISchemaElement[] => (
  elements.filter((element) => (
    name !== (element as FieldElement).options?.inputMeta?.name
  ))
);

export const getUIElementWithName = (
  name: string,
  elements: UISchemaElement[],
): UISchemaElement | undefined => (
  elements.find((element) => (
    name === (element as FieldElement).options?.inputMeta?.name
  ))
);
