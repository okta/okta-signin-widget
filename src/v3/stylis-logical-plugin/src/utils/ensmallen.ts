import { compile, serialize, stringify } from 'stylis';

export const ensmallen = (css: string) => {
  return serialize(compile(css), stringify);
};
