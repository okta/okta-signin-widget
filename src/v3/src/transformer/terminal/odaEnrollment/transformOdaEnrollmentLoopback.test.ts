/*
 * Copyright (c) 2022-present, Okta, Inc. and/or its affiliates. All rights reserved.
 * The Okta software accompanied by this notice is provided pursuant to the Apache License, Version 2.0 (the "License.")
 *
 * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0.
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *
 * See the License for the specific language governing permissions and limitations under the License.
 */

import { IdxContext, IdxStatus } from '@okta/okta-auth-js';

import { getStubFormBag, getStubTransaction } from '../../../mocks/utils/utils';
import {
  DescriptionElement,
  ListElement,
  TitleElement,
  UISchemaElement,
  UISchemaLayout,
  WidgetProps,
} from '../../../types';
import { transformOdaEnrollmentLoopback } from './transformOdaEnrollmentLoopback';

describe('Terminal ODA enrollment Loopback transformer', () => {
  const transaction = getStubTransaction(IdxStatus.TERMINAL);
  const formBag = getStubFormBag();
  let widgetProps: WidgetProps;

  beforeEach(() => {
    formBag.uischema.elements = [];
    transaction.messages = [];
    widgetProps = {};
  });

  it('adds iOS UI elements appropriately when platform is iOS', () => {
    transaction.context = {
      deviceEnrollment: {
        value: {
          name: 'oda',
          challengeMethod: 'LOOPBACK',
          signInUrl: 'https://okta.com',
          platform: 'IOS',
        },
      },
    } as unknown as IdxContext;

    const updatedFormBag = transformOdaEnrollmentLoopback({ formBag, transaction, widgetProps });

    expect(updatedFormBag).toMatchSnapshot();

    expect(formBag.uischema.elements.length).toBe(3);

    expect(formBag.uischema.elements[0].type).toBe('Title');
    expect((formBag.uischema.elements[0] as TitleElement).options.content).toBe('enroll.title.oda');

    expect((formBag.uischema.elements[2] as UISchemaElement).type).toBe('List');

    const listElement = formBag.uischema.elements[2] as ListElement;
    const listItems = listElement.options.items;

    expect(listItems.length).toBe(6);

    const listItemOne = listItems[0];
    expect((listItemOne as UISchemaLayout).elements.length).toBe(2);
    expect(((listItemOne as UISchemaLayout).elements[0] as DescriptionElement).options.content).toBe('enroll.mdm.step.copyLink');

    const listItemFour = listItems[3];
    expect(((listItemFour as UISchemaLayout).elements[0] as DescriptionElement).options.content).toBe('enroll.oda.step1');

    const listItemFive = listItems[4];
    expect(((listItemFive as UISchemaLayout).elements[1] as DescriptionElement).options.content).toBe('https://okta.com');
  });

  it('adds iOS UI elements appropriately when platform is Android and challenge method is loopback', () => {
    transaction.context = {
      deviceEnrollment: {
        value: {
          name: 'oda',
          challengeMethod: 'LOOPBACK',
          signInUrl: 'https://okta.com',
          platform: 'ANDROID',
        },
      },
    } as unknown as IdxContext;

    const updatedFormBag = transformOdaEnrollmentLoopback({ formBag, transaction, widgetProps });

    expect(updatedFormBag).toMatchSnapshot();

    expect(formBag.uischema.elements.length).toBe(3);

    expect(formBag.uischema.elements[0].type).toBe('Title');
    expect((formBag.uischema.elements[0] as TitleElement).options.content).toBe('enroll.title.oda');

    expect((formBag.uischema.elements[2] as UISchemaElement).type).toBe('List');

    const listElement = formBag.uischema.elements[2] as ListElement;
    const listItems = listElement.options.items;

    expect(listItems.length).toBe(4);

    const listItemOne = listItems[0];
    expect(((listItemOne as UISchemaLayout).elements[0] as DescriptionElement).options.content).toBe('enroll.oda.android.step1');

    const listItemThree = listItems[2];
    expect(((listItemThree as UISchemaLayout).elements[1] as DescriptionElement).options.content).toBe('https://okta.com');
  });
});
