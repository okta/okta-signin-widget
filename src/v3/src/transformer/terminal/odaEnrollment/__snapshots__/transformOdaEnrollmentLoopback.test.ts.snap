// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Terminal ODA enrollment Loopback transformer adds iOS UI elements appropriately when platform is Android and challenge method is loopback 1`] = `
Object {
  "data": Object {},
  "dataSchema": Object {
    "fieldsToExclude": [Function],
    "fieldsToTrim": Array [],
    "fieldsToValidate": Array [],
    "submit": Object {
      "step": "challenge-authenticator",
    },
  },
  "schema": Object {},
  "uischema": Object {
    "elements": Array [
      Object {
        "options": Object {
          "content": "enroll.title.oda",
        },
        "type": "Title",
      },
      Object {
        "contentType": "subtitle",
        "options": Object {
          "content": "enroll.explanation.p1",
        },
        "type": "Description",
      },
      Object {
        "noMargin": true,
        "options": Object {
          "items": Array [
            Object {
              "elements": Array [
                Object {
                  "noMargin": true,
                  "options": Object {
                    "content": "enroll.oda.android.step1",
                  },
                  "type": "Description",
                },
              ],
              "type": "VerticalLayout",
            },
            Object {
              "elements": Array [
                Object {
                  "noMargin": true,
                  "options": Object {
                    "content": "enroll.oda.step1",
                  },
                  "type": "Description",
                },
              ],
              "type": "VerticalLayout",
            },
            Object {
              "elements": Array [
                Object {
                  "noMargin": true,
                  "options": Object {
                    "content": "enroll.oda.step2",
                  },
                  "type": "Description",
                },
                Object {
                  "noMargin": true,
                  "options": Object {
                    "content": "https://okta.com",
                  },
                  "type": "Description",
                },
                Object {
                  "label": "enroll.oda.org.copyLink",
                  "options": Object {
                    "onClick": [Function],
                    "step": "",
                    "type": "button",
                    "variant": "secondary",
                  },
                  "type": "Button",
                },
              ],
              "type": "VerticalLayout",
            },
            Object {
              "elements": Array [
                Object {
                  "noMargin": true,
                  "options": Object {
                    "content": "enroll.oda.step6",
                  },
                  "type": "Description",
                },
              ],
              "type": "VerticalLayout",
            },
          ],
          "type": "ordered",
        },
        "type": "List",
      },
    ],
    "type": "VerticalLayout",
  },
}
`;

exports[`Terminal ODA enrollment Loopback transformer adds iOS UI elements appropriately when platform is iOS 1`] = `
Object {
  "data": Object {},
  "dataSchema": Object {
    "fieldsToExclude": [Function],
    "fieldsToTrim": Array [],
    "fieldsToValidate": Array [],
    "submit": Object {
      "step": "challenge-authenticator",
    },
  },
  "schema": Object {},
  "uischema": Object {
    "elements": Array [
      Object {
        "options": Object {
          "content": "enroll.title.oda",
        },
        "type": "Title",
      },
      Object {
        "contentType": "subtitle",
        "options": Object {
          "content": "enroll.explanation.p1",
        },
        "type": "Description",
      },
      Object {
        "noMargin": true,
        "options": Object {
          "items": Array [
            Object {
              "elements": Array [
                Object {
                  "noMargin": true,
                  "options": Object {
                    "content": "enroll.mdm.step.copyLink",
                  },
                  "type": "Description",
                },
                Object {
                  "label": "enroll.mdm.copyLink",
                  "options": Object {
                    "onClick": [Function],
                    "step": "",
                    "type": "button",
                    "variant": "secondary",
                  },
                  "type": "Button",
                },
              ],
              "type": "VerticalLayout",
            },
            Object {
              "elements": Array [
                Object {
                  "noMargin": true,
                  "options": Object {
                    "content": "enroll.mdm.step.pasteLink",
                  },
                  "type": "Description",
                },
              ],
              "type": "VerticalLayout",
            },
            Object {
              "elements": Array [
                Object {
                  "noMargin": true,
                  "options": Object {
                    "content": "enroll.oda.step3",
                  },
                  "type": "Description",
                },
              ],
              "type": "VerticalLayout",
            },
            Object {
              "elements": Array [
                Object {
                  "noMargin": true,
                  "options": Object {
                    "content": "enroll.oda.step1",
                  },
                  "type": "Description",
                },
              ],
              "type": "VerticalLayout",
            },
            Object {
              "elements": Array [
                Object {
                  "noMargin": true,
                  "options": Object {
                    "content": "enroll.oda.step2",
                  },
                  "type": "Description",
                },
                Object {
                  "noMargin": true,
                  "options": Object {
                    "content": "https://okta.com",
                  },
                  "type": "Description",
                },
                Object {
                  "label": "enroll.oda.org.copyLink",
                  "options": Object {
                    "onClick": [Function],
                    "step": "",
                    "type": "button",
                    "variant": "secondary",
                  },
                  "type": "Button",
                },
              ],
              "type": "VerticalLayout",
            },
            Object {
              "elements": Array [
                Object {
                  "noMargin": true,
                  "options": Object {
                    "content": "enroll.oda.step6",
                  },
                  "type": "Description",
                },
              ],
              "type": "VerticalLayout",
            },
          ],
          "type": "ordered",
        },
        "type": "List",
      },
    ],
    "type": "VerticalLayout",
  },
}
`;
